### PR TITLE
[Binance] Add subAccount Transfert History

### PR DESCRIPTION
**Pull request by trixprod**
Add two funding record api fetch,
one for the master account ( with a mandatory sub account email ),
one for the subaccount ( nothing mandatory ).

### CODE REVIEW

File: xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  public static final class TransfertHistory {
+    private String from;
+    private String to;
+    private String asset;
+    private BigDecimal qty;
+    private long time;
+
+    public String getFrom() {
+      return from;
+    }
```
Comment (walec51) [inline xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java]: if your using lombok then these setters / getters are not needed

File: xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java [hunk 2]
```diff
@@ -287,6 +304,52 @@ public TradeHistoryParams createFundingHistoryParams() {
                 });
       }
 
+      final String finalEmail = email;
```
Comment (walec51) [inline xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java]: I don't think this is needed

File: xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java [hunk 3]
```diff
@@ -287,6 +304,52 @@ public TradeHistoryParams createFundingHistoryParams() {
...
+
+      if (email != null) {
+        super.getTransfertHistory(email, startTime, endTime, page, limit)
+            .forEach(
+                a -> {
+                  result.add(
+                      new FundingRecord(
```
Comment (walec51) [inline xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java]: please use builder

File: xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java [hunk 4]
```diff
@@ -287,6 +304,52 @@ public TradeHistoryParams createFundingHistoryParams() {
...
+      }
+
+      if (subAccount) {
+
+        Integer type = deposits && withdrawals ? null : deposits ? 1 : 0;
+        super.getSubUserHistory(asset, type, startTime, endTime, limit)
+            .forEach(
+                a -> {
+                  result.add(
+                      new FundingRecord(
```
Comment (walec51) [inline xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java]: please use builder

File: xchange-core/src/main/java/org/knowm/xchange/dto/account/FundingRecord.java [hunk 5]
```diff
@@ -298,7 +298,13 @@ public String toString() {
      * outflows defined above or their nature could not have been deduced from the exchanges
      * response
      */
-    OTHER_OUTFLOW(false);
+    OTHER_OUTFLOW(false),
+
+    /** Used for inflows that come from a sub account to the master account */
+    SUBACCOUNT_INFLOW(true),
```
Comment (walec51) [inline xchange-core/src/main/java/org/knowm/xchange/dto/account/FundingRecord.java]: I think more generic names would be better as this is to Binance-specific

some exchanges have a flat structure with many account for a given currency and no master

I would suggest:

```
    /**
     * Withdrawal from the exchange a blockchain address or a bank account.
     *
     * Usually results in a withdrawal fee being subtracted
     */
    WITHDRAWAL(false),

    /** Deposit from a blockchain address or a bank account to the exchange */
    DEPOSIT(true),

    /** Used for transfers between exchanges accounts */
    INTERNAL_WITHDRAWAL(false);

    /** Used for transfers between exchanges accounts */
    INTERNAL_DEPOSIT(true),

    /** Used for transfers between exchanges accounts */
    INTERNAL_WITHDRAWAL(false);
```

File: xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java (hunk 1 above)
Comment (trixprod) [reply xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java]: I haven't created unit test but if you remove getter, are you still able to create UT ?

File: xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java (hunk 2 above)
Comment (trixprod) [reply xchange-binance/src/main/java/org/knowm/xchange/binance/service/BinanceAccountService.java]: Lambda need final field. email field is null then affected to a value if needed.

File: xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java (hunk 1 above)
Comment (walec51) [reply xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java]: I'm not telling to remove getters
I'm telling to remove the getter code you have written by hand
lombok will generate getters for you - this is how the `@Data` annotation you have used here works https://projectlombok.org/features/Data

Comment (Spyros-Stylianou) [reply xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java]: file name has a typo

Comment (trixprod) [reply xchange-binance/src/main/java/org/knowm/xchange/binance/dto/account/TransfertHistoryResponse.java]: > file name has a typo

Fixed all references from Transfert to Transfer ( without the 't' )

**Pull request merged**